src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1153,13 +1176,19 @@ public StateChange ForkProject(
     {
         var projectId = newProjectState.Id;
 
-        Contract.ThrowIfFalse(ProjectStates.ContainsKey(projectId));
-        var newStateMap = ProjectStates.SetItem(projectId, newProjectState);
+        Contract.ThrowIfFalse(ContainsProject(projectId));
+
+        using var _ = ArrayBuilder<ProjectState>.GetInstance(ProjectStates.Length, out var newProjectStates);
+        foreach (var state in ProjectStates)
+            newProjectStates.Add(state == oldProjectState ? newProjectState : state);
+
+        // Sort this before realizing the ImmutableArray to avoid an extra array allocation for sorting during the branch.
+        newProjectStates.Sort(CompareProjectStates);
```

Comment:
Won't this list already be sorted? As long as we have a strong guarantee that the ID doesn't change (which I assume we always had), this could just be implemented via a ImmutableArray.SetItem on the single entry that needs to change. Since that distills down to an internal Array.Copy I presume the runtime can optimize that quite nicely.

Reply:
Yeah, I guess assuming oldProjectState.Id == newProjectState.Id is probably a valid thing to do and will simplify this code.